providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -513,135 +487,122 @@ static const OSSL_PARAM *mlkem_gen_settable_params(ossl_unused void *genctx,
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
 
-    mkey->keytype = MLKEM_KEY_TYPE_768;
+    genok = seed != NULL ?
+        ossl_mlkem_vgenkey_seed(v, seed, seedlen, nopub, 0, &key->prvkey,
+                                key->ctx) :
+        ossl_mlkem_vgenkey_rand(v, NULL,       0, nopub, 0, &key->prvkey,
+                                key->ctx);
 
-    if (mkey->encoded_pubkey == NULL) {
-        mkey->encoded_pubkey = OPENSSL_malloc(OSSL_MLKEM768_PUBLIC_KEY_BYTES);
-        if (mkey->encoded_pubkey == NULL)
-            goto err;
-    }
+    /* Single-use seeds */
+    if (seed != NULL)
+        OPENSSL_cleanse(seed, MLKEM_SEED_BYTES);
+    gctx->seeded = 0;
 
-    if (mkey->encoded_privkey == NULL) {
-        mkey->encoded_privkey = OPENSSL_malloc(OSSL_MLKEM768_PRIVATE_KEY_BYTES);
-        if (mkey->encoded_privkey == NULL)
-            goto err;
-    }
+    if (genok)
+        return key;
 
-    if (gctx->seed != NULL) {
-        debug_print("MLKEMKM generate keys from seed");
-        if (!ossl_mlkem768_private_key_from_seed(&mkey->privkey, gctx->seed,
-                                                 OSSL_MLKEM_SEED_BYTES,
-                                                 mkey->mlkem_ctx)
-            || !ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
-                                                  &mkey->privkey)
-            || !ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->privkey)
-            || !ossl_mlkem768_marshal_public_key(mkey->encoded_pubkey,
-                                                 &mkey->pubkey))
-            goto err;
-    } else {
-        debug_print("MLKEMKM generate random keys");
-        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
-                                        &mkey->privkey, mkey->mlkem_ctx)
-            || !ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
-                                                  &mkey->privkey)
-            || !ossl_mlkem768_public_from_private(&mkey->pubkey,
-                                                  &mkey->privkey))
-            goto err;
-    }
-
-    debug_print("MLKEMKM gen returns set %p\n", mkey);
-    return mkey;
-
-err:
-    OPENSSL_free(mkey);
+    ossl_mlkem_vcleanse_prvkey(v, &key->prvkey);
+    OPENSSL_free(key);
     return NULL;
 }
 
-static void mlkem_gen_cleanup(void *genctx)
+static void mlkem_gen_cleanup(void *vgctx)
 {
-    struct mlkem_gen_ctx *gctx = genctx;
-
-    OPENSSL_free(gctx->seed);
-    debug_print("MLKEMKM gen cleanup for %p\n", gctx);
-    OPENSSL_free(gctx);
+    OPENSSL_free(vgctx);
 }
 
-static void *mlkem_dup(const void *vsrckey, int selection)
+static void *mlkem_dup(const void *vkey, int selection)
 {
-    const MLKEM768_KEY *srckey = (const MLKEM768_KEY *)vsrckey;
-    MLKEM768_KEY *dstkey;
+    const MLKEM_KEY *key = (const MLKEM_KEY *)vkey;
+    vinfo_t v = key->vinfo;
```

Comment:
Worthwhile defensive check for vkey != NULL before dereferencing it?

Reply:
I don't think it is possible to attempt to dup a NULL key.